system/ui/text.py

```diff
@@ -0,0 +1,66 @@
+import sys
+import pyray as rl
+
+from openpilot.system.hardware import HARDWARE
+from openpilot.system.ui.lib.button import gui_button
+from openpilot.system.ui.lib.scroll_panel import GuiScrollPanel
+from openpilot.system.ui.lib.application import gui_app
+
+MARGIN = 50
+SPACING = 50
+FONT_SIZE = 60
+LINE_HEIGHT = 64
+BUTTON_SIZE = rl.Vector2(310, 160)
+
+DEMO_TEXT = """This is a sample text that will be wrapped and scrolled if necessary.
+            The text is long enough to demonstrate scrolling and word wrapping.""" * 20
+
+def wrap_text(text, font_size, max_width):
+  lines = []
+  current_line = ""
+  font = gui_app.font()
+
+  for word in text.split():
+    test_line = current_line + word + " "
+    if rl.measure_text_ex(font, test_line, font_size, 0).x <= max_width:
+      current_line = test_line
+    else:
+      lines.append(current_line)
+      current_line = word + " "
+  if current_line:
+    lines.append(current_line)
+
+  return lines
+
+
+def main():
+  gui_app.init_window("Text")
+
+  text_content = sys.argv[1] if len(sys.argv) > 1 else DEMO_TEXT
+
+  textarea_rect = rl.Rectangle(MARGIN, MARGIN, gui_app.width - MARGIN * 2, gui_app.height - MARGIN * 2 - BUTTON_SIZE.y - SPACING)
+  wrapped_lines = wrap_text(text_content, FONT_SIZE, textarea_rect.width - 20)
+  content_rect = rl.Rectangle(0, 0, textarea_rect.width - 20, len(wrapped_lines) * LINE_HEIGHT)
+  scroll_panel = GuiScrollPanel(textarea_rect, content_rect, show_vertical_scroll_bar=True)
+
+  while not rl.window_should_close():
+    rl.begin_drawing()
+    rl.clear_background(rl.BLACK)
+
+    scroll = scroll_panel.handle_scroll()
+
+    rl.begin_scissor_mode(int(textarea_rect.x), int(textarea_rect.y), int(textarea_rect.width), int(textarea_rect.height))
+    for i, line in enumerate(wrapped_lines):
+      position = rl.Vector2(textarea_rect.x + scroll.x, textarea_rect.y + scroll.y + i * LINE_HEIGHT)
+      rl.draw_text_ex(gui_app.font(), line.strip(), position, FONT_SIZE, 0, rl.WHITE)
+    rl.end_scissor_mode()
```

Comment:
the text area tears while scrolling

Reply:
@deanlee
[set FLAG_VSYNC_HINT to avoid tearing while scrolling](https://github.com/commaai/openpilot/pull/34583/commits/c274b79922f995b75755cd54b7bee7e679609a36)